tests/components/numato/test_binary_sensor.py

```diff
@@ -0,0 +1,70 @@
+"""Tests for the numato binary_sensor platform."""
+from homeassistant.helpers import discovery
+from homeassistant.setup import async_setup_component
+
+from .common import NUMATO_CFG, mockup_raise
+
+MOCKUP_ENTITY_IDS = {
+    "binary_sensor.numato_binary_sensor_mock_port2",
+    "binary_sensor.numato_binary_sensor_mock_port3",
+    "binary_sensor.numato_binary_sensor_mock_port4",
+}
+
+
+async def test_failing_setups_no_entities(hass, numato_fixture, monkeypatch):
+    """When port setup fails, no entity shall be created."""
+    monkeypatch.setattr(numato_fixture.NumatoDeviceMock, "setup", mockup_raise)
+    assert await async_setup_component(hass, "numato", NUMATO_CFG)
+    await hass.async_block_till_done()
+    for entity_id in MOCKUP_ENTITY_IDS:
+        assert entity_id not in hass.states.async_entity_ids()
+
+
+async def test_setup_callbacks(hass, numato_fixture, monkeypatch):
+    """During setup a callback shall be registered."""
+
+    numato_fixture.discover()
+
+    def mock_add_event_detect(self, port, callback, direction):
+        assert self == numato_fixture.devices[0]
+        assert port == 1
+        assert callback is callable
+        assert direction == numato_fixture.BOTH
+
+    monkeypatch.setattr(
+        numato_fixture.devices[0], "add_event_detect", mock_add_event_detect
+    )
+    assert await async_setup_component(hass, "numato", NUMATO_CFG)
+
+
+async def test_hass_binary_sensor_notification(hass, numato_fixture):
+    """Test regular operations from within Home Assistant."""
+    assert await async_setup_component(hass, "numato", NUMATO_CFG)
+    await hass.async_block_till_done()  # wait until services are registered
+
+    # TODO: Understand why creating a task makes below assert work and remove this workaround
+    async def async_anything():
+        pass
+
+    await hass.async_create_task(async_anything())
```

Comment:
Do we get the same result with:
```py
await asyncio.sleep(0)
```
?

Maybe it's a question of needing to yield to the event loop again to let other tasks finish?

Reply:
I think I know now. The callback is a sync function and needs to schedule work threadsafe onto the event loop. So there are more tasks created. The callback should be run in the executor thread pool.

So use `hass.async_add_executor_job` to call the callback and await until that is done.